### PR TITLE
New style tile

### PR DESCRIPTION
**Pull request by hvy**
New style tile function to support double backprop.

### CODE REVIEW

File: chainer/functions/array/tile.py [hunk 1]
```diff
@@ -15,22 +17,23 @@ def __init__(self, reps):
...
-            raise ValueError('all elements in reps must be zero or larger')
+            raise ValueError('All elements in reps must be zero or larger')
 
     def check_type_forward(self, in_types):
         type_check.expect(in_types.size() == 1)
 
     def forward(self, inputs):
-        self.retain_inputs(())
         self._in_shape = inputs[0].shape
         self._in_dtype = inputs[0].dtype
```
Comment (kmaehashi) [inline chainer/functions/array/tile.py]: This line seems unneeded.

File: chainer/functions/array/tile.py [hunk 2]
```diff
@@ -41,21 +44,23 @@ def backward(self, inputs, grads):
         elif ndim < len(reps):
             shape = (1,) * (len(reps) - ndim) + shape
 
-        if grads[0].shape == ():
+        gy, = grad_outputs
+        if gy.shape == ():
             # This case should be treated differently because numpy.num would
```
Comment (kmaehashi) [inline chainer/functions/array/tile.py]: This comment is no longer valid as `numpy.sum` (seems like a typo) is now removed.

**Comment by hvy** [comment]
@kmaehashi Thanks for the comments. I'll take a look!

File: chainer/functions/array/tile.py (hunk 1 above)
Comment (hvy) [reply chainer/functions/array/tile.py]: I think you commented on the old code. It has been removed!

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
@kmaehashi Removed the unnecessary gy shape check before rebasing to master. It passes the tests. Could you check the code if it looks okay?

File: chainer/functions/array/tile.py (hunk 1 above)
Comment (kmaehashi) [reply chainer/functions/array/tile.py]: Ah, both old code and new code does not using `_in_dtype`.

Comment (hvy) [reply chainer/functions/array/tile.py]: Thanks, I'll fix this one.

**Comment by hvy** [comment]
Rebased on top of master.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
LGTM! I will merge this after CI passing.

**Pull request merged**